nixos/hostapd: rewrite to support multi-AP, WiFi 6, password from file, and more

oddlama opened the pull request:
###### Description of changes

I've noticed that the hostapd module is quite outdated and was in need of some love. This PR tries to make the hostapd module ready for modern AP setups, WiFi 6, multi-AP functionality and other things outlined below. Later I noticed that #49171 existed but was recently closed as dead, but this PR implements all of that and more.

At this point this is basically a full rewrite of this module, which is a breaking change and was necessary to properly expose the useful parts of hostapd's config. The necessary changes for users of this module are very limited (move to correct interface attrset, refactor use of wpaPassword -> saePasswords).

Since this is my first contribution I'm unsure whether a renaming notice or something like that would be required or preferred in that case. I'm happy to receive any suggestions update this PR accordingly.

The notable module changes are:

- `hostapd` is now started with additional systemd sandbox/hardening options
- A single-daemon can now manage multiple distinct interfaces, which is why all configuration had to be moved into `services.hostapd.interfaces` (similar to what #49171 tried to accomplish)
- By default WPA3-SAE will be used, but WPA2 and WPA3-SAE-TRANSITION are supported, too
- Added `passwordFile`-like options for WPA and SAE which closes #24288.
- Structure new configuration by major wifi versions (WiFi 5,6,7) and expose relevant options in the module.
- Add options to support MAC ACL allow/deny lists
- Include proper documentation for each option.
- Removes `with lib;` usage as per #208242.

I considered implementing this in the `settings`-style which seems to be preferred, but due to hostap's very unconventional config format (ini-like but sometimes order-dependent, without categories and special value formats) I opted against for now. It seems like an unproportional amount of work to make sure it stays compatible with their config format.

A second commit enables new stable features such as WiFi6 and structures the .config for the hostapd package.
These changes are important to support modern APs configurations and WiFi 6. Short overview:

- CONFIG_IEEE80211AX support (WiFi6)
- CONFIG_SAE_PK (pubkey authenticated WPA3, important for open APs)
- CONFIG_DRIVER_NONE (standalone RADIUS server)
- CONFIG_OCV (Operating Channel Validation)
- Enable epoll on linux systems
- Remove deprecated TKIP support
- Fix misspelling (CONFIG_INTERNETWORKING != CONFIG_INTERWORKING)
- The .config was restructured into sections to reflect the
  upstream defconfig order and for easier updating in the future

The third and last commit enables WPA3-SAE-PK in wpa_supplicant, which is currently missing.
For completeness, CONFIG_IEEE80211AX is also added to the list of enabled options.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

oddlama committed:
wpa_supplicant: enable WPA3-SAE-PK

oddlama committed:
hostapd: enable new stable features such as WiFi6 and structure .config

These changes are important to support modern APs configurations.
Short overview:

- CONFIG_IEEE80211AX support (WiFi6)
- CONFIG_SAE_PK (pubkey authenticated WPA3)
- CONFIG_DRIVER_NONE (standalone RADIUS server)
- CONFIG_OCV (Operating Channel Validation)
- Enable epoll on linux systems
- Remove deprecated TKIP support
- Fix misspelling (CONFIG_INTERNETWORKING != CONFIG_INTERWORKING)
- The .config was restructured into sections to reflect the
  upstream defconfig order and for easier updating in the future

oddlama committed:
nixos/hostapd: rewrite to support multi-AP, password from file, and more

At this point this is basically a full rewrite of this module, which
is a breaking change and was necessary to properly expose the useful
parts of hostapd's config. The notable changes are:

- `hostapd` is now started with additional systemd sandbox/hardening options
- A single-daemon can now manage multiple distinct radios and BSSs, which is
  why all configuration had to be moved into `hostapd.radios`
- By default WPA3-SAE will be used, but WPA2 and WPA3-SAE-TRANSITION are
  supported, too
- Added passwordFile-like options for wpa and sae
- Add new relevant options for MAC ACL, WiFi5, WiFi6 and WiFi7 configuration
- Implements RFC42 as far as reasonable for hostapd
- Removes `with lib;`

RaitoBezarius commented on pkgs/os-specific/linux/hostapd/default.nix:
```diff
@@ -1,5 +1,7 @@
 { lib, stdenv, fetchurl, pkg-config, libnl, openssl, sqlite ? null }:
 
+let
```
empty let

mweinelt reviewed (changes requested):
I did a first pass to try to get to know the broad strokes, because it's a lot to take in.

I'm not sure the package changes are going into the correct direction, in fact the package should probably reduce the config options to those that actually differ from upstream, instead of piling on more and more options.

Adding hardening to the systemd unit is a good idea, and I added some remarks about things I think we can or should refactor.

Finally, my worries about the module as a whole is that it is a whopping 1000 lines long now. It exposes a lot of options explicitly, which will likely require a lot of maintenance going forward. This is the reason for RFC42 and `settings` approach, in that it allows to specify a limited set of important option, while at same time providing an escape hatch by allowing custom key/value pairs at any place.

Instead of your helper functions (L28+), you could use `apply` in `mkOption` to integrate the mapping logic more tightly.

oddlama commented on pkgs/os-specific/linux/hostapd/default.nix:
Again, I can eliminate duplicates if this is the way you want it to be. I initially made it explicit because options cannot be disabled by specifying `CONFIG_ABC=n` (it mostly counts as yes if they are defined in any way) thus it felt cleaner to me.

The existing config also had duplicates (CONFIG_IEEE80211AC=y, CONFIG_IEEE80211N=y), and invalid entries (CONFIG_INTERNETWORKING=y), so I just threw it away and started anew (maybe too rash).

oddlama commented on pkgs/os-specific/linux/hostapd/default.nix:
```diff
@@ -38,28 +56,59 @@ stdenv.mkDerivation rec {
+    CONFIG_OWE=y
+    CONFIG_OCV=y
+
+    # TKIP is considered insecure and upstream support will be removed in the future
+    CONFIG_NO_TKIP=y
```
I'm not sure if I am missing something but I think that's incorrect. This is an option to disable TKIP and it is commented out in the [defconfig for 2.10](https://w1.fi/cgit/hostap/tree/hostapd/defconfig?h=hostap_2_10#n397), so it's enabled unless there another mechanism at play that disables TKIP.

oddlama commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1007 @@
+      serviceConfig = {
+        ExecStart = "${pkgs.hostapd}/bin/hostapd ${concatStringsSep " " runtimeConfigFiles}";
+        Restart = "always";
+        ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
```
I tested it with kill and it works as expected. Would `hostapd_cli reconfigure` be preferable?

oddlama commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1007 @@
+          ##### User-provided extra configuration ##########################################
+
+          EOF
+          cat ${escapeShellArg (pkgs.writeText ifcfg.extraConfig)} >> "$hostapd_config_file"
```
I'm not sure I fully understand what you mean, what would the benefit be for having the extraConfig in /etc?

oddlama commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1007 @@
+
+    environment.systemPackages = [pkgs.hostapd];
+
+    services.udev.packages = optionals (any (i: i.countryCode != null) (attrValues cfg.interfaces)) [pkgs.crda];
```
Fully agreed, I just tried to not change the semantics of the original module. Will change.

oddlama commented:
> I'm not sure the package changes are going into the correct direction, in fact the package should probably reduce the config options to those that actually differ from upstream, instead of piling on more and more options.

In that case I can of course rebase the config for wpa_supplicant and hostapd on their current upstream defaults. Altoutgh this might have to be done periodically (e.g. when CONFIG_IEEE80211AX will become a default in the next years) to keep a clean diff.

> Finally, my worries about the module as a whole is that it is a whopping 1000 lines long now. It exposes a lot of options explicitly, which will likely require a lot of maintenance going forward. This is the reason for RFC42 and `settings` approach, in that it allows to specify a limited set of important option, while at same time providing an escape hatch by allowing custom key/value pairs at any place.

I very briefly addressed that in the PR description, the issue is that the hostapd config is parsed by a custom C parser. The format is mostly key=value pairs, but doesn't follow this rule strictly. I opted against using settings style for now because there are several caveats that I don't know how to work around. For example:

- `sae_password` is an option that can be repeated
- The whole config is order dependent. If you write `bss=wlan0_a` you enter a new context which allows most (not all) options to be respecified for that sub-configuration
- 
There are several cases of such inconveniences, and I have no idea how that could be mapped to RFC 42 without severely limiting generality. For now I just exposed some of the available options which are actually useful for AP setup.

> Instead of your helper functions (L28+), you could use `apply` in `mkOption` to integrate the mapping logic more tightly.

Done.

RaitoBezarius commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1011 @@
+      ${optionalString (ifcfg.authentication.wpaPskFile != null) ''
+        wpa_passphrase=${ifcfg.authentication.wpaPskFile}
+      ''}
+      ${optionalString (ifcfg.authentication.saePasswords != []) (concatMapStrings (pw: "sae_password=${pw}\n") ifcfg.authentication.saePasswords)}
```
this cannot work as `pw` is not necessarily a string :P

oddlama replied:
damn I missed that when promoting saePasswords to a set, thanks. Will push an update in a minute.

oddlama commented:
I've added preliminary support for multiple BSS definitions per radio. I decided to do this since it impossible to do that manually using `extraConfig` and in order to prevent another breaking change in the future. It is working, but I'm still ironing out some details and thus I put this back into draft mode for now. I'll update here once finished.

GoogleBot42 commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1160 @@
+
+          ##### User-provided extra radio configuration ##########################################
+          EOF
+          cat ${escapeShellArg (pkgs.writeText radioCfg.extraConfig)} >> "$hostapd_config_file"
```
I think `writeText` needs a filename from my testing.  So something like

```
cat ${escapeShellArg (pkgs.writeText "filename" radioCfg.extraConfig)} >> "$hostapd_config_file"
```

GoogleBot42 commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1160 @@
+
+                ##### User-provided extra BSS configuration ##########################################
+                EOF
+                cat ${escapeShellArg (pkgs.writeText bssCfg.extraConfig)} >> "$hostapd_config_file"
```
Same as other comment

GoogleBot42 commented:
Awesome work!  I wish I discovered this a big sooner before I delved into hostapd.

OpenWRT has some pretty nice auto configure scripts to automatically set the capabilities and calculates the center frequency.  Without those extra values openwrt calculates and adds to the hostapd config, I wasn't able to get high speeds.  I did make an ugly script that amends the hostapd configuration to add these capabilities by quering the hardware at service start but it is incomplete and definitely not ready to show off. I think a cool follow up to this PR might be to add some auto configure magic. :)

I do have a question about the configuration format.

```nix
{
    services.hostapd = {
      enable = true;
      radios = {
        wlan1 = {
          hwMode = "a";
          countryCode = "US";
          networks.wlan1 = {
            ssid = "tower";
            authentication.saePasswords = [{ password = "pass"; }]; 
          };
        };
      };
    };
}
```

It seems a bit odd to me that `wlan1` is specified twice.  Hostapd's config format is... terrible so maybe there's a need for it but otherwise I see maybe an alternative or two.  Some other ideas that don't repeat the wireless interface.

```nix
{
    services.hostapd = {
      enable = true;
      radios = {
        wlan1 = {
          hwMode = "a";
          countryCode = "US";
          networks = [  # use an array
            {
            ssid = "tower";
            authentication.saePasswords = [{ password = "pass"; }]; 
            }
          ];
        };
      };
    };
}
```

```nix
{
    services.hostapd = {
      enable = true;
      radios = {
        wlan1 = {
          hwMode = "a";
          countryCode = "US";
          networks.tower = { # the SSID is the key
            authentication.saePasswords = [{ password = "pass"; }]; 
          };
        };
      };
    };
}
```

mweinelt commented:
> It seems a bit odd to me that wlan1 is specified twice

The first occurence specifies the radio, the second on the bss. It looks that way, because hostapd requires `interrface=` for the first bss, but `bss=` for every following bss.


```
          networks = [  # use an array
```

We decided against that, because values below arrays are terrible to reference.

oddlama commented:
>           networks.tower = { # the SSID is the key;  I assume all networks *must* have an SSID right?

Adding to what @mweinelt commented, indexing by SSID is also not possible since two BSS can technically advertise under the same SSID (but different BSSID). Indexing by BSSID instead is also not trivial, since the BSSID may be derived automatically by hostapd and/or the device driver, so you don't want to be required to specify that. Maybe this can be made easier by using artificial (meaningless) keys instead, but I'm not sure yet. The main issue is that hostapd maintains a weird connection between logical interfaces names and bss defintions. (Suggestions are welcome)

> OpenWRT has some pretty nice auto configure scripts to automatically set the capabilities and calculates the center frequency. 

In this PR I was mainly focusing on making the module cover hostapd's abilities. There definitely are things we could learn from what OpenWRT does, so that may be something to investigate for the future. :)

> So wifi6 usage doesn't have to wait for upstream to release a new version maybe this line:
> 
>     require_he=${bool01 radioCfg.wifi6.require}
>
> Can be replaced with:
>
>     ${optionalString radioCfg.wifi6.require "require_he=1"}

Sure, probably better that way. I'll also change some other options that are not strictly required to be included each time.

GoogleBot42 commented:
I'm having trouble getting multiple SSIDs working.  I'm probably just doing something wrong due to a lack of understanding.  It seems to not like `wlan1-1`.  Openwrt is able to do it on my 5Ghz interface so I don't think it's a hardware issue.

<details>
<summary>My (simplified) config</summary>

```nix
{
  services.hostapd = {
    enable = true;
    radios = {
      # 2.4GHz
      wlp4s0 = {
        hwMode = "g";
        networks = {
          wlp4s0 = {
            ssid = "AP 1";
            authentication.saePasswordsFile = "path/to/file";
          };
          wlp4s0-1 = {
            ssid = "AP 2";
            authentication.saePasswordsFile = "path/to/file2";
          };
        };
      };

      # 5GHz
      wlan1 = {
        hwMode = "a";
        networks = {
          wlan1 = {
            ssid = "AP 1";
            authentication.saePasswordsFile = "path/to/file";
          };
          wlan1-1 = {
            ssid = "AP 3";
            authentication.saePasswordsFile = "path/to/file2";
          };
        };
      };
    };
  };
}
```

</details>

<details>
<summary>Hostapd logs</summary>

```
Apr 04 20:22:21 router hostapd[15239]: Could not read interface wlan1-1 flags: No such device
Apr 04 20:22:21 router hostapd[15239]: nl80211: Driver does not support authentication/association or connect commands
Apr 04 20:22:21 router hostapd[15239]: nl80211: deinit ifname=wlan1-1 disabled_11b_rates=0
Apr 04 20:22:21 router hostapd[15239]: Could not read interface wlan1-1 flags: No such device
Apr 04 20:22:21 router hostapd[15239]: nl80211 driver initialization failed.
Apr 04 20:22:21 router hostapd[15239]: wlan1-1: interface state UNINITIALIZED->DISABLED
Apr 04 20:22:21 router hostapd[15239]: wlan1-1: AP-DISABLED
Apr 04 20:22:21 router hostapd[15239]: wlan1-1: CTRL-EVENT-TERMINATING
Apr 04 20:22:21 router hostapd[15239]: hostapd_free_hapd_data: Interface wlan1-1 wasn't started
Apr 04 20:22:21 router hostapd[15239]: wlp4s0-1: interface state UNINITIALIZED->DISABLED
Apr 04 20:22:21 router hostapd[15239]: wlp4s0-1: AP-DISABLED
Apr 04 20:22:21 router hostapd[15239]: wlp4s0-1: CTRL-EVENT-TERMINATING
Apr 04 20:22:21 router hostapd[15239]: hostapd_free_hapd_data: Interface wlp4s0-1 wasn't started
Apr 04 20:22:21 router systemd[1]: hostapd.service: Main process exited, code=exited, status=1/FAILURE
```

</details>

oddlama commented:
> I'm having trouble getting multiple SSIDs working. I'm probably just doing something wrong due to a lack of understanding. It seems to not like `wlan1-1`. Openwrt is able to do it on my 5Ghz interface so I don't think it's a hardware issue.

I think this is to be expected right now, I'm still in the process of testing which options hostapd additionally wants to have in such cases. Some adapters seem to require you to set the bssid manually, while others can use the driver to automatically derive that. Would you be able and willing to share the hostapd config that openwrt generated for your device? It would be immensely helpful to get another datapoint there.

oddlama commented:
I've changed the meaning of `macAcl` values to match user expectations (similar to OpenWRT), where "allow" means use an "allowlist" not the reversed meaning as implied by hostapd's setting.

GoogleBot42 commented:
Sorry was a bit busy.  I booted up openwrt.  This is the config it generated for the 5Ghz radio.  I can provide the 2.4Ghz radio too if needed.

Almost all of the config is unrelated I'm sure but I don't want to accidentally remove something important.

<details>

<summary>Config</summary>

```
driver=nl80211
logger_syslog=127
logger_syslog_level=2
logger_stdout=127
logger_stdout_level=2
country_code=US
ieee80211d=1
ieee80211h=1
hw_mode=a
beacon_int=100
channel=128
chanlist=128

tx_queue_data2_burst=2.0
ieee80211n=1
ht_coex=0
ht_capab=[HT40-][LDPC][SHORT-GI-20][SHORT-GI-40][TX-STBC][RX-STBC1][MAX-AMSDU-7935]
ieee80211ac=1
vht_oper_chwidth=2
vht_oper_centr_freq_seg0_idx=114
vht_capab=[RXLDPC][SHORT-GI-80][SHORT-GI-160][TX-STBC-2BY1][SU-BEAMFORMER][SU-BEAMFORMEE][MU-BEAMFORMER][MU-BEAMFORMEE][RX-ANTENNA-PATTERN][TX-ANTENNA-PATTERN][RX-STBC-1][SOUNDING-DIMENSION-2][BF-ANTENNA-4][VHT160][MAX-MPDU-11454][MAX-A-MPDU-LEN-EXP7]
ieee80211ax=1
he_oper_chwidth=2
he_oper_centr_freq_seg0_idx=114
he_su_beamformer=1
he_mu_beamformer=1
he_default_pe_duration=4
he_rts_threshold=1023
he_mu_edca_qos_info_param_count=0
he_mu_edca_qos_info_q_ack=0
he_mu_edca_qos_info_queue_request=0
he_mu_edca_qos_info_txop_request=0
he_mu_edca_ac_be_aifsn=8
he_mu_edca_ac_be_aci=0
he_mu_edca_ac_be_ecwmin=9
he_mu_edca_ac_be_ecwmax=10
he_mu_edca_ac_be_timer=255
he_mu_edca_ac_bk_aifsn=15
he_mu_edca_ac_bk_aci=1
he_mu_edca_ac_bk_ecwmin=9
he_mu_edca_ac_bk_ecwmax=10
he_mu_edca_ac_bk_timer=255
he_mu_edca_ac_vi_ecwmin=5
he_mu_edca_ac_vi_ecwmax=7
he_mu_edca_ac_vi_aifsn=5
he_mu_edca_ac_vi_aci=2
he_mu_edca_ac_vi_timer=255
he_mu_edca_ac_vo_aifsn=5
he_mu_edca_ac_vo_aci=3
he_mu_edca_ac_vo_ecwmin=5
he_mu_edca_ac_vo_ecwmax=7
he_mu_edca_ac_vo_timer=255

radio_config_id=b6303943e81f4d060feb5336474e8530
interface=wlan1
ctrl_interface=/var/run/hostapd
ap_isolate=1
bss_load_update_period=60
chan_util_avg_period=600
disassoc_low_ack=1
skip_inactivity_poll=0
preamble=1
wmm_enabled=1
ignore_broadcast_ssid=0
uapsd_advertisement_enabled=1
utf8_ssid=1
multi_ap=0
wpa_passphrase=XXXX
wpa_psk_file=/var/run/hostapd-wlan1.psk
auth_algs=1
wpa=2
wpa_pairwise=CCMP
ssid=SSID-1
bridge=br-lan
wds_bridge=
snoop_iface=br-lan
wpa_disable_eapol_key_retries=0
wpa_key_mgmt=WPA-PSK
okc=0
disable_pmksa_caching=1
dynamic_vlan=0
vlan_naming=1
vlan_no_bridge=1
vlan_file=/var/run/hostapd-wlan1.vlan
qos_map_set=0,0,2,16,1,1,255,255,18,22,24,38,40,40,44,46,48,56
config_id=0cb8625cf6aa257abb3be95a0fb495d1
bssid=00:0a:52:08:38:33


bss=wlan1-1
ctrl_interface=/var/run/hostapd
ap_isolate=1
bss_load_update_period=60
chan_util_avg_period=600
disassoc_low_ack=1
skip_inactivity_poll=0
preamble=1
wmm_enabled=1
ignore_broadcast_ssid=0
uapsd_advertisement_enabled=1
utf8_ssid=1
multi_ap=0
wpa_passphrase=XXXX
wpa_psk_file=/var/run/hostapd-wlan1-1.psk
auth_algs=1
wpa=2
wpa_pairwise=CCMP
ssid=SSID-1
bridge=br-lan
wds_bridge=
snoop_iface=br-lan
wpa_disable_eapol_key_retries=0
wpa_key_mgmt=WPA-PSK
okc=0
disable_pmksa_caching=1
dynamic_vlan=0
vlan_naming=1
vlan_no_bridge=1
vlan_file=/var/run/hostapd-wlan1-1.vlan
qos_map_set=0,0,2,16,1,1,255,255,18,22,24,38,40,40,44,46,48,56
config_id=1a342509404a31d8624b19719017c755
bssid=02:0a:52:08:38:33
```

</details>

esclear commented:
This PR seems very promising and desirable.
@oddlama, is there anything I can do to help with this PR?

RaitoBezarius commented:
> This PR is 95% finished, the main part that is still missing is the bssid calculation logic. That is where we should probably do something similar to what OpenWrt does, but last time I looked at it seemed non-trivial. Once that is sorted out this should be ready to go. Apart from that, general feedback and code review is always welcome.

Actually, I believe we should cut on the scope and do a follow up for BSSID, don't you think so?

oddlama commented:
> Actually, I believe we should cut on the scope and do a follow up for BSSID, don't you think so?

Yeah, I just looked at it again and you are absolutely right. This is going to take quite some effort to implement correctly. To make multiple BSS usable right now I'll just expose the bssid option directly for the time being and cut it there. Actually let me do that now.

oddlama commented:
So I'd consider this done for now, all the main features are ready and working on the cards I have access to. More testing would be greatly appreciated, as always. Changes in the most recent push:

- Added assertions to ensure valid BSS names
- Expose bssid and require setting it when using multiple BSS
- Rebased on master

A followup PR might later address stuff like

- bssid deduction for multiple bss
- Exposing more WiFi 6/6E related options

RaitoBezarius commented:
@mweinelt thought I was saying this was good for a merge in the next days, but just to be clear: this should go *after* the branch-off for 23.05 obviously, it has high potential of breakage and it's too late.

esclear commented:
I am seeing next to no debug output, even when configuring a network with `logLevel = 0;`.

oddlama commented:
> I am seeing next to no debug output, even when configuring a network with `logLevel = 0;`.

That's just hostapd unfortunately. There are almost no debug messages present in it.

NinjaTrappeur commented:
This changes drastically the scope of tihs NixOS module. We now need to track upstream configuration file options directly in here. That's fine, I don't want to block that change provided people are willing to maintain that on the long run. 

I personally do not have the bandwidth to do so. I only use this module on my home access point. Wifi card that I refresh twice a decade at most.

I'm afraid it'd be hard to move back to something more barebones when we'll adopt this without breaking live setups.

@oddlama do you plan to maintain that on the long run? If so, could you consider adding yourself as a maintainer of this NixOS module?

oddlama commented:
> This changes drastically the scope of tihs NixOS module. We now need to track upstream configuration file options directly in here. That's fine, I don't want to block that change provided people are willing to maintain that on the long run.

It does. But we only really need to track the options that have now been integrated into this module. I will do that of course, but history suggests for better or for worse that the existing options will never be touched in a meaningful way ever again by hostapd. Other then that, there is no inherent need to track future options per se - so probably just some minor convenience updates to support WiFi 6E and WiFi 7 without requiring `extraConfig`.

> I personally do not have the bandwidth to do so. I only use this module on my home access point. Wifi card that I refresh twice a decade at most.

No need to :) I'm mainly interested in general feedback
 
> I'm afraid it'd be hard to move back to something more barebones when we'll adopt this without breaking live setups if we end up lacking ~man~people-power at some point (like we already do).

All I can say is that I do plan to maintain this myself as good as I can.

RaitoBezarius commented:
I am also willing to join the maintenance.

rnhmjoj commented:
> bssid deduction for multiple bss

Or, hear me out, generate random BSSIDs each time the service start, similarly to what we do for IPv6 addresses.
This prevents your AP from being (trivially) tracked by location services such as [this](https://support.google.com/maps/answer/1725632?hl=en) or [this](https://security.stackexchange.com/questions/95426/how-to-protect-a-wifi-network-from-microsoft-wifi-sense) without having to name your network `myssid_optout_nomap`, which is ridiculous.

rnhmjoj commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1190 @@
-      };
+          For each radio, hostapd requires a separate logical interface (like wlp3s0, wlp3s1, ...).
+          A default interface is usually be created automatically by your system, but to use
+          multiple radios with a single device, you will likely have to use {command}`iw` to create
```
I believe there's `networking.wlanInterfaces` for this. It would be better to recommend the declarative approach.

oddlama replied:
Good call, that's probably a good thing to do.

rnhmjoj commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1190 @@
+                  message = "hostapd radio ${radio} bss ${bss}: The bss (network) name ${bss} is invalid. It must be prefixed by the radio name for reasons internal to hostapd. A valid name would be e.g. ${radio}, ${radio}-1, ...";
+                }
+                {
+                  assertion = (length (attrNames radioCfg.networks) > 1) -> (auth.bssid != null);
```
```suggestion
                  assertion = (length (attrNames radioCfg.networks) > 1) -> (bssCfg.bssid != null);
```

oddlama replied:
Thanks, I missed that. Done

oddlama commented:
Rebased and moved release notes to 23.11

rnhmjoj commented:
Can you fix `nixosTests.wpa_supplicant`? It would also be interesting to test SAE and mixed WPA2-SAE now that `services.hostapd` supports them.

oddlama commented:
Tests are now ready. 3 SSIDs are created with different authentication schemes, but currently only wpa3-sae is tested. I don't see a way to tell `networking.wireless` that we want to connect to 3 networks simultaneously using 3 different interfaces. We could theoretically run 3 machines instead, but I'm not sure if that's desirable. Please add your thoughts :)

RaitoBezarius commented:
> Tests are now ready. 3 SSIDs are created with different authentication schemes, but currently only wpa3-sae is tested. I don't see a way to tell `networking.wireless` that we want to connect to 3 networks simultaneously using 3 different interfaces. We could theoretically run 3 machines instead, but I'm not sure if that's desirable. Please add your thoughts :)

I see multiple things:

- having 3 nodes connecting simulatenously to an AP is not shocking to me
- having 3 test variants doing the same test but connecting to one of each network
- driving networkmanager/something to connect to the 3 nets using 3 different interfaces on the same node

You can pick one depending what is the rationale for the system under test, I don't know that much hostapd to decide for a sensible one.

oddlama commented:
> You can pick one depending what is the rationale for the system under test, I don't know that much hostapd to decide for a sensible one.

Thanks for the pointers :) I've decided to go with your second suggestion, since this allows isolated testing and it's easier to determine whether the connection is established correctly. Now there are 5 isolated tests:

- basic wpa_supplicant tests (preexisting)
- 4 tests wpa_supplicant connecting to hostapd
  - to a SAE only hotspot using SAE
  - to a SAE-WPA2 mixed hotspot using SAE
  - to a SAE-WPA2 mixed hotspot using WPA2
  - to a WPA2 only hotspot using WPA2

rnhmjoj commented:
Hostapd documentation is inexistent and I don't have the logs anymore, but I recall that if you specify the channel width with `vht_oper_chwidth` you also need to (re)compute the `vht_oper_centr_freq_seg0_idx` and `vht_oper_centr_freq_seg1_idx`.
It doesn't look like you can do that from this module, though: does this make `operatingChannelWidth` useless?

rnhmjoj commented:
> I mainly cut the features at that point to not make the PR even bigger. 

Sure, that's fine. I just missed `radioCfg.extraConfig` because it's not used in `writeRadioBaseConfig`.

rnhmjoj commented:
For me the PR is pretty much good to go: it's a vast improvement over the original module, which honestly was kind of a toy. I would recommend two things:

- Try to make the resulting hostapd.conf more human-readable: it sometimes happen that you need to inspect it and it's kind of a mess of blank lines, right now. You can look at 33372e8b75 (wpa_supplicant) for inspiration. In this case indentation is not required, so it's a bit easier to do.

-  Add `mkChangedOptionModule` for the removed options: a blind `nixos-rebuild` should at least warn that a breaking change took place and point to the release notes or (better) offer migration instructions.

Also the relase notes need to be moved to 23.11, 23.05 is already out.

oddlama commented:
Good points, I'll have a look at the linked commit regarding blank lines and add some `mkChangedOptionModule`. But I do believe that I already moved the release notes to 23.11 last time I pushed.

Although I will probably have to update them again to to new conflicts (the 23.11 notes were still quite empty when I moved them).

rnhmjoj commented:
> Although I will probably have to update them again to to new conflicts

Oh yeah, I confused it for 23.05.

oddlama commented:
Alright, I've applied the final changes you suggested:

- The configuration files are now generated without unnecessary whitespace (I left some intentional line breaks for grouping options into related categories).
- I've used `mkRemovedOptionModule` on all renamed options with a explanatory message instead of relying on `mkChangedOptionModule`, since the latter unfortunately cannot be used for automatic rewriting in this complex usecase (for example `services.hostapd.ssid` would have to be rewritten as `services.hostapd.radios.${oldConfig.services.hostapd.interface}.networks.${oldConfig.services.hostapd.interface}.ssid` so the new attribute path is dependent on an old option, which isn't supported). So using `mkRemovedOptionModule` with a good message seemed like the best alternative.

Would be good to go from my side.

rnhmjoj reviewed (approved):
Great, thank you!

rnhmjoj reviewed (approved):
Great, thank you!

rnhmjoj commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1228 @@
+        "wpa_passphrase=${bssCfg.authentication.wpaPassword}"
+      ++ optional (bssCfg.authentication.wpaPskFile != null)
+        "wpa_passphrase=${bssCfg.authentication.wpaPskFile}"
+      ++ (map genSaePasswordEntry bssCfg.authentication.saePasswords)
```
I think you're missing a newline here, in case both wpaPasswords and `wpaPskFile` are null:
```suggestion
      ++ (map genSaePasswordEntry bssCfg.authentication.saePasswords)
      ++ optional (bssCfg.authentication.wpaPasswordFile != null ||
                   bssCfg.authentication.saePasswordsFile != null) ""
```

rnhmjoj replied:
No, you'll get an invalid configuration file. I'm just not sure where it's best to put the newline: either here or in the script that concatenates the files.

oddlama replied:
Could this be an more general issue because the file doesn't end with a newline (because of `concatStringsSep "\n"`)? If so, this should fail in every case and I should probably use `concatStringsMap (x: "${x}\n")`.

2xsaiko commented:
> The whole config is order dependent. If you write bss=wlan0_a you enter a new context which allows most (not all) options to be respecified for that sub-configuration

I just saw this PR now after writing my own module for hostapd, which does use the settings attrset style (though I call it config right now, should rename it). What I do in this case is just sort the keys based on a custom order which has bss at the top among other things before emitting them to the file.

Here's my module: https://git.dblsaiko.net/systems/tree/configurations/turris/hostapd.nix?h=turris&id=7dfcd8ff6e928078c485ff2934aba6277f54f872

And example usage: https://git.dblsaiko.net/systems/tree/configurations/turris/wifi.nix?h=turris&id=7dfcd8ff6e928078c485ff2934aba6277f54f872

oddlama commented:
> What I do in this case is just sort the keys based on a custom order which has bss at the top among other things before emitting them to the file.

Thanks for your suggestion. Unfortunately I personally think that there is no feasible path forward to a settings-style configuration for hostapd.

The main issue being this [gigantic list of options](https://w1.fi/cgit/hostap/tree/hostapd/config_file.c?id=4120f9d7ab12660364ca4c9dd6e6e4cb66f5342a#n2310) which we would have to manually sift through to determine whether an option is global or BSS-specific and whether it is repeatable. And we'd have to repeat that on every hostapd release. So I believe the small benefits don't justify the immense maintaining cost here (at least I wouldn't want to do that each release). Since hostapd's format is proprietary it's not really possible to design something that doesn't suddenly break when hostapd introduces a new obscure option. Missing a sorting rule or repeated option would be way worse for users than resorting to `extraConfig`, especially since users can write their own config adapter if they want.

This PR already exposes first-class options for everything needed by the average user. a future PR might extend this module to add more support for WiFi 6 related configuration. So `extraConfig` usage should already be rare (or simple) and thus I think having `extraConfig` is currently preferrable.

The proper way to solve this would be to push hostapd to use a proper standardized configuration format. Not sure if that will ever happen though.

2xsaiko commented:
> The main issue being this [gigantic list of options](https://w1.fi/cgit/hostap/tree/hostapd/config_file.c?id=4120f9d7ab12660364ca4c9dd6e6e4cb66f5342a#n2310) which we would have to manually sift through to determine whether an option is global or BSS-specific and whether it is repeatable. And we'd have to repeat that on every hostapd release.

Why? I'm not familiar with hostapd's options but from what I can figure the only thing whose position actually matters is the position of the bss setting in relation to the other settings (the other sorting I do is pretty much just so the final config file is more organized). This can be handled by dividing it into the existing structural options you already have.

You don't have to explicitly list every possible parameter, it can just be general enough so that every option can be specified, and doesn't have to forbid invalid settings (not to mention that extraConfig has the exact same problems, but on top of that doesn't allow you to use any of the nice NixOS option stuff like mkDefault).

That parser seems straight forward enough, with each line being key=value, and multiple occurrences being theoretically possible. If we can map that onto nixpkgs options, imo it would be good enough, more extensible, and easier to use, and I see no downsides except for the very unlikely case that hostapd decides to add an option into the config syntax that isn't key=value. Even then, special casing for single options (v/ht_capab would be a candidate) could in theory be added separately, like is done with services.dnsmasq.settings.server (in fact, looking at [dnsmasq](https://github.com/NixOS/nixpkgs/blob/nixos-23.05/nixos/modules/services/networking/dnsmasq.nix), there's even pkgs.formats.keyValue which looks like it would handle most of the generation for hostapd as well).

oddlama commented:
The point is that maintaining exceptions is very impractical for maintenance, not that it isn't possible in theory. Someone has to do it on every release, and I'm not very keen on being that person. The way the module is structured now allows updates without changes to the module. That being said, your `pkgs.formats.keyValue` suggestion would be a valid way to replace `extraConfig` with something more usable, without going full RFC42.

> Why? I'm not familiar with hostapd's options but from what I can figure the only thing whose position actually matters is the position of the bss setting in relation to the other settings (the other sorting I do is pretty much just so the final config file is more organized).

`sae_passwords` is one example where order matters (due to last-match logic), and it was introduced not too long ago. So assuming this might happen again is not so far-fetched.

> You don't have to explicitly list every possible parameter, it can just be general enough so that every option can be specified, and doesn't have to forbid invalid settings (not to mention that extraConfig has the exact same problems, but on top of that doesn't allow you to use any of the nice NixOS option stuff like mkDefault).

For proper RFC42 you really would have to do that. hostapd has too many options that take specially formatted lists (some with duplicate key mechanics, some without) and we'd ideally want to disallow using global options in bss-config segments and vice-versa.

Just replacing `extraConfig` by `pkgs.formats.keyValue` with `listsAsDuplicateKeys = true` is definitely an option that I would consider. It would solve your `mkDefault` problem, but doesn't introduce additional maintenance cost. Let's wait one or two days for other people to voice their opinions.

Tungsten842 commented on pkgs/os-specific/linux/wpa_supplicant/default.nix:
```diff
@@ -48,6 +48,7 @@ stdenv.mkDerivation rec {
     CONFIG_HS20=y
     CONFIG_HT_OVERRIDES=y
     CONFIG_IEEE80211AC=y
+    CONFIG_IEEE80211AX=y
```
Both CONFIG_IEEE80211AX=y and CONFIG_SAE=y
are already enabled by default
https://w1.fi/cgit/hostap/tree/wpa_supplicant/defconfig#n487
https://w1.fi/cgit/hostap/tree/wpa_supplicant/defconfig#n255

oddlama replied:
You are right, but there's a lot more unnecessary assignments already in there (e.g. CONFIG_IEEE80211AC=y is also default), mostly because the upstream `defconfig` changes from time to time. I don't see an issue with explicitly including stuff like this if just to document that it was being though of.

2xsaiko replied:
> CONFIG_IEEE80211AX=y

This one isn't in release 2.10, looks like CONFIG_SAE is though.

Tungsten842 replied:
I think that we shouldn't introduce more unnecessary assignments, if we don't do that this list is going to get bigger and bigger until one day we are going to maintain an entire defconfig. I also think that we should remove every duplicated assignments in another pr.

oddlama replied:
@Tungsten842  I just re-checked and saw that your original your comment is wrong. You link to the wpa_supplicant defconfig, which is different from the hostapd defconfig. In hostapd almost nothing is enabled by default.

Tungsten842 replied:
> @Tungsten842 I just re-checked and saw that your original your comment is wrong. You link to the wpa_supplicant defconfig, which is different from the hostapd defconfig. In hostapd almost nothing is enabled by default.

I linked both of them.

oddlama replied:
No, you linked to CONFIG_SAE and CONFIG_IEEE80211AX in [tree/wpa_supplicant](https://w1.fi/cgit/hostap/tree/). In hostapd the option is not enabled by default, [even on master](https://w1.fi/cgit/hostap/tree/hostapd/defconfig#n159)

oddlama replied:
Ah sorry, my mistake. This is about wpa_supplicant. Forget what I said.

oddlama replied:
Sorry again for the confusion. The redundant definitions have been removed.

Tungsten842 commented on pkgs/os-specific/linux/hostapd/default.nix:
```diff
@@ -38,28 +56,54 @@ stdenv.mkDerivation rec {
+    CONFIG_NO_TKIP=y
+
+    # Misc
+    CONFIG_PKCS12=y
```
CONFIG_PKCS12=y
CONFIG_IPV6=y
CONFIG_RSN_PREAUTH=y
CONFIG_DPP=y
CONFIG_DPP2=y
are already enabled by default:
https://w1.fi/cgit/hostap/tree/hostapd/defconfig
Also CONFIG_OWE, CONFIG_WNM are experimental

oddlama replied:
There seems to be a more fundamental question on how we want to handle this. Do we only want to track differences from upstream defconfig, or do we want to explicitly list all options that were considered?

>Also CONFIG_OWE, CONFIG_WNM are experimental
CONFIG_OWE is necessary for WPA3 compatibility, we shouldn't turn that off. Disabling WNM by default is a good idea because it is an incomplete implementation.

2xsaiko replied:
> CONFIG_IPV6=y
> CONFIG_RSN_PREAUTH=y

These aren't in 2.10: https://w1.fi/cgit/hostap/tree/wpa_supplicant/defconfig?h=hostap_2_10

oddlama replied:
I think rebasing on 2.10 is a fine choice.

oddlama replied:
Done.

oddlama commented:
Rebased the hostapd package configuration on the current defconfig.

oddlama commented on pkgs/os-specific/linux/wpa_supplicant/default.nix:
Rebased on current defconfig. Now we only track changes.

oddlama commented:
Based on @2xsaiko's suggestion I've replaced `extraConfig` with `extraSettings` to make this as much compatible with RFC 42 as reasonably possible without introducing future maintenance concerns.

oddlama commented:
> And if you want to reduce maintenance cost, I suggest removing some of those options that are only used to add a single value to the file, like apIsolate, and have users write settings.ap_isolate = true instead. Of course, the trade-off is documentation.

I've put quite some effort into the documentation strings, and since the hostapd documentation is basically nonexistent I'd like to keep them so users will have an easier time configuring this. But this shouldn't conflict with moving them into the freeform settings and removing the manual assignments.

> Is there an actual _concrete case_ where this currently won't work? I would argue any potential option that may be added in the future which doesn't conform to the generated format isn't enough to justify not fully using settings for everything right now and adapting it when it happens, because it seems very unlikely. The module has to be changed either way when hostapd updates and someone wants to use that option.

The BSS state of course, but we have to handle that one specially either way. I'm not seeing any others right now, I'm probably just too paranoid. So maybe two `settings` definitions is the way to go afterall, one for interface-global settings and one for bss-specific settings. Lets try :)

2xsaiko commented:
> I've put quite some effort into the documentation strings, and since the hostapd documentation is basically nonexistent I'd like to keep them so users will have an easier time configuring this. But this shouldn't conflict with moving them into the freeform settings and removing the manual assignments.

Yeah, it's probably good to keep it, it's well written and the upstream documentation is meh (no manpage and the only real description of the options I've found is the example config file in the source code).

> > Freeform attributes cannot depend on other attributes of the same set without infinite recursion:
> > To prevent this, declare options for all attributes that need to depend on others. For above example this means to declare logLevel to be an option.
> 
> So apparently we won't be able to skip defining the options we want to set anyway.

Is that an issue? Which ones would you want to set depending on another value in settings? The only thing that setting a value there should do (inside the same module) should be to include it in the configuration file, imo, and not cause other options to change.

I'd keep those where you actually want to do that, like you have right now, outside of settings, and do something like

    settings.isolate_ap = mkDefault cfg.isolateAp;

for each (using mkDefault, mkIf, etc. where applicable).

oddlama commented:
Updated to use RFC42 settings style options (after handling radios and networks of course).

> I'd keep those where you actually want to do that, like you have right now, outside of settings, and do something like
> 
> ```
> settings.isolate_ap = mkDefault cfg.isolateAp;
> ```
> 
> for each (using mkDefault, mkIf, etc. where applicable).

I've only used `mkDefault` for values that have no first-class option from this module (like `logger_syslog` for example), so that users will notice when they override a value that is already explicitly set by the module.

2xsaiko commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1235 @@
+      preStart = concatStringsSep "\n" (mapAttrsToList makeRadioRuntimeFiles cfg.radios);
+
+      serviceConfig = {
+        ExecStart = "${pkgs.hostapd}/bin/hostapd ${concatStringsSep " " runtimeConfigFiles}";
```
Is it maybe a good idea to add a services.hostapd.package option so the hostapd version can be swapped out without a nixpkgs overlay?

oddlama replied:
Certainly good to have, will add.

dblsaiko reviewed (commented):
Awesome! I really like how this looks (and I hope I haven't annoyed you too much asking you to change all this stuff)

Another thing I'd like to see is being able to customize config generation at runtime, like what you're doing right now for e.g. passwordFile but in an externally extensible way. This would allow NixOS configurations (or other modules) to add something such as automatically detecting ht_capab based on what the adapter says (something that OpenWrt does and I potentially wanted to try getting to work in NixOS), without having to use a nixpkgs fork.

Should work by adding an option under each radio and each network that then gets put together into your makeRadioRuntimeFiles script, and then using those options to implement passwordFile and the rest of what you currently have under makeRadioRuntimeFiles.

oddlama commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1235 @@
+                  };
+
+                  ssid = mkOption {
+                    default = config.system.nixos.distroId;
```
Good point. And since the `ssid` is usually the first thing that users will change, I'd opt for removing the default all together.

oddlama commented:
Thanks for the suggestions, I've now changed the following things:

- Introduced a package option to be able to override the hostapd package
- Removed default definitions for SSIDs
- Moved all the runtime config generation logic to a new `dynamicConfigScripts` attrset, where new scripts can be added if necessary. One set per radio and one per network BSS of course. The module pre-fills some bss specific stuff to generate the mac allow/deny lists and to add passwords from files.

> Awesome! I really like how this looks (and I hope I haven't annoyed you too much asking you to change all this stuff)

Not at all, I'm glad for the feedback! I believe this way the PR has come a long way from the initial draft.

> Another thing I'd like to see is being able to customize config generation at runtime, like what you're doing right now for e.g. passwordFile but in an externally extensible way. This would allow NixOS configurations (or other modules) to add something such as automatically detecting ht_capab based on what the adapter says (something that OpenWrt does and I potentially wanted to try getting to work in NixOS), without having to use a nixpkgs fork.

I've also thought about that, but after seeing the amount of edge-case handling in OpenWrt I dismissed the thought for now. If you manage to do that generically this would be awesome to have in NixOS.

2xsaiko commented:
> I've also thought about that, but after seeing the amount of edge-case handling in OpenWrt I dismissed the thought for now. If you manage to do that generically this would be awesome to have in NixOS.

Yeeeah, when I first took a look I kinda bailed out when seeing the two ~1.5k line shell scripts to generate this stuff, using lots of OpenWrt specific programs and other scripts. I'd still like to do it at some point though because it would be really cool to not have to worry about adapter configuration (for drivers that behave).

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2311.section.md:
```diff
@@ -18,6 +18,15 @@
+  - `hostapd` is now started with additional systemd sandbox/hardening options for better security.
+  - `services.hostapd.interface` was replaced with a per-radio and per-bss configuration scheme using [services.hostapd.radios](#opt-services.hostapd.radios).
+  - `services.hostapd.wpa` has been replaced by [services.hostapd.radios.&lt;name&gt;.networks.&lt;name&gt;.authentication.wpaPassword](#opt-services.hostapd.radios._name_.networks._name_.authentication.wpaPassword) and [services.hostapd.radios.&lt;name&gt;.networks.&lt;name&gt;.authentication.saePasswords](#opt-services.hostapd.radios._name_.networks._name_.authentication.saePasswords) which con ...
+  - The default authentication has been changed to WPA3-SAE. Options for other (legacy) schemes are stil availabe.
```
We can't just tell it to use the best supported scheme?

oddlama replied:
No. You can of course choose wpa3 transition if you want (allowing both sae and wpa2 if the client doesn't support sae3) but this of course allows for downgrade attacks. Since almost all devices can do sae today just fine, it should be the default.

SuperSandro2000 commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1283 @@
+# to make informed decisions you will probably need to look
+# at hostapd's code. You have been warned, proceed with care.
+let
+  inherit
```
Why do we maintain this list of lib functions?

rnhmjoj replied:
Yeah, I'd say to just use `lib.` everywhere, it's just three letters anyway.

oddlama replied:
Is this a relevant change though?

Since some consider [`with lib` to be an antipattern](https://nix.dev/recipes/best-practices#with-scopes) I chose this to make the code more readable without having `lib.` bloating the line everywhere. Since serveral files in nixpkgs do the same thing I didn't expect this to be controversial.

RaitoBezarius replied:
Let's leave those nitpicks out of the PR as it's already quite hairy, complicated and probably taking a lot of patience from the author. :)

SuperSandro2000 commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1283 @@
-    logger_stdout=-1
-    logger_stdout_level=${toString cfg.logLevel}
+      hostapd_config_file=/run/hostapd/${escapeShellArg radio}.hostapd.conf
+      rm -f "$hostapd_config_file"
```
Do we want to reset permissions here?

oddlama replied:
Shouldn't make a difference, this is an ephemeral file managed by the module either way. Or do you have something specific in mind?

SuperSandro2000 commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1283 @@
-          interface to provide IP addresses to the associated stations, and
-          NAT (from the wireless interface to an upstream interface).
-        '';
+      package = mkOption {
```
Use mkPackageOption

oddlama replied:
Will do.

SuperSandro2000 commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1283 @@
-          Most applications will probably use the default.
+            # Legacy WPA2 example
+            wlp4s0 = {
+              # countryCode = "US";
```
I always thought you set this on the AP only.

SuperSandro2000 commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1283 @@
+              example = "US";
+              type = types.nullOr types.str;
+              description = mdDoc ''
+                Country code (ISO/IEC 3166-1). Used to set regulatory domain.
```
Do we need to load the regulatory database under the hardware options for this to work?

oddlama replied:
`crda` is added to the system packages which should suffice.

SuperSandro2000 commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1283 @@
+              type = types.int;
+              description = mdDoc ''
+                The channel to operate on. Use 0 to enable ACS (Automatic Channel Selection).
+                Beware that not every device supports ACS in which case {command}`hostapd`
```
How common is that? If the devices are in the wast minority which do not support this we might want to set it to 0 by default

mweinelt replied:
Very uncommon.

oddlama replied:
Some devices support it, but the vast majority of kernel drivers does *not* implement it. You want to use it when you can, but you'll likely only be able to use it on very well supported chipsets. `mt7921` based cards works for example.

SuperSandro2000 commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1283 @@
+        cfg.radios));
 
-    environment.systemPackages =  [ pkgs.hostapd ];
+    environment.systemPackages = [cfg.package];
```
```suggestion
    environment.systemPackages = [ cfg.package ];
```

oddlama replied:
Sorry this is very opinionated and I don't think this is relevant as long as there isn't an agreed upon style in nixpkgs. Any future formatter would easily change this. Until then, I'll just use alejandra to format my PRs.

SuperSandro2000 commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1283 @@
+    environment.systemPackages = [cfg.package];
 
-    services.udev.packages = optionals (cfg.countryCode != null) [ pkgs.crda ];
+    services.udev.packages = with pkgs; [crda];
```
```suggestion
    services.udev.packages = with pkgs; [ crda ];
```

oddlama replied:
Same thing here.

SuperSandro2000 commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1007 @@
+        DevicePolicy = "closed";
+        DeviceAllow = "/dev/rfkill rw";
+        NoNewPrivileges = true;
+        PrivateUsers = false; # hostapd requires true root access.
```
I imagine this to be a pain to debug since you don't get any meaningful errors if some capability is missing

oddlama replied:
Yeah, I'd postpone this for now. It's non trivial (to me) to figure out exactly what hostapd could possibly want. Missing something would break it, so better change it when we are sure about what it requires.

RaitoBezarius commented:
@oddlama Can you fix the conflicts please?

oddlama commented:
> @oddlama Can you fix the conflicts please?

Done :)

RaitoBezarius commented on nixos/modules/services/networking/hostapd.nix:
```diff
@@ -1,222 +1,1279 @@
+
+              capabilities = mkOption {
+                type = types.listOf types.str;
+                default = ["HT40" "HT40-" "SHORT-GI-20" "SHORT-GI-40"];
```
HT40- is unsupported with ACS, an assert would be nice :).

oddlama replied:
It's not implemented in hostapd, but in theory there's nothing preventing them from doing ACS on HT40- right? They seem to have a [TODO in there](https://w1.fi/cgit/hostap/tree/src/ap/acs.c#n1025). Do we still want to assert that anyway?

oddlama replied:
From matrix: We do.

oddlama commented:
Latest commit changes the RFC42 settings such that they show an error when redefining non-list symbols with different values, instead of silently making a list of them.

oddlama commented:
Replaced `hwMode` in favor of `band` to make configuration experience better (we follow OpenWRT here). Also added assertion that ensures HT40- is not used with ACS (unsupported by hostapd).

rnhmjoj commented on nixos/doc/manual/release-notes/rl-2311.section.md:
```diff
@@ -20,6 +20,15 @@
 
 - [Apache Guacamole](https://guacamole.apache.org/), a cross-platform, clientless remote desktop gateway. Available as [services.guacamole-server](#opt-services.guacamole-server.enable) and [services.guacamole-client](#opt-services.guacamole-client.enable) services.
 
+- Support for WiFi6 (IEEE 802.11ax) and WPA3-SAE-PK was enabled in the `hostapd` package.
```
This line should go under "highlights" and the rest either in "backward incompatibilities" or "notable changes".
"New services" seems the wrong section.

RaitoBezarius reviewed (approved):
This PR is one of the most excellent work I have seen in nixpkgs, @oddlama has been very patient, and I feel like it's time to move on to the next PR for more improvements in that area.

I will merge this now, feel free to revert me if that's too premature in your opinion.

jcumming commented:
Wow! Thanks for all of the work @oddlama !

This is a huge improvement over the massive string of opaque hostapd options I currently use for multi-ap.
